### PR TITLE
IGNITE-26584 Discovery optimizations for MultiDC

### PR DESCRIPTION
**Pull request by anton-vinogradov**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/MdcAwareNodesComparator.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+/** Compares nodes using the Data Center id as a primary factor. */
+public class MdcAwareNodesComparator implements Comparator<TcpDiscoveryNode> {
+    /** */
+    @Override public int compare(TcpDiscoveryNode n1, TcpDiscoveryNode n2) {
+        String n1DcId = n1.dataCenterId() == null ? "" : n1.dataCenterId();
+        String n2DcId = n2.dataCenterId() == null ? "" : n2.dataCenterId();
+
+        int res = n1DcId.compareTo(n2DcId);
+
+        if (res == 0) {
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/MdcAwareNodesComparator.java]: Redundant braces

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/TcpDiscoveryNodesRing.java [hunk 2]
```diff
@@ -506,7 +506,10 @@ public void clear() {
             if (filtered.size() < 2)
                 return null;
 
-            Iterator<TcpDiscoveryNode> iter = filtered.iterator();
+            NavigableSet<TcpDiscoveryNode> sorted = new TreeSet<>(new MdcAwareNodesComparator());
+            sorted.addAll(filtered);
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/TcpDiscoveryNodesRing.java]: I'm not sure if `TreeSet` always preserves order of added elements if all elements are equal according to comparator, so it potentianly can break something in non-multi-DC cluster. Let's use sorting only for multi-DC environment. Also it's a potentially dangerous feature even for multi-DC, I think it's worth to have a flag to disable it explicitely.

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/TcpDiscoveryNodesRing.java [hunk 3]
```diff
@@ -506,7 +506,10 @@ public void clear() {
             if (filtered.size() < 2)
                 return null;
 
-            Iterator<TcpDiscoveryNode> iter = filtered.iterator();
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/TcpDiscoveryNodesRing.java:509]: Methodd `TcpDiscoverySpi#forceNextNodeFailure` -> `ServerImpl#forceNextNodeFailure` depend on `nextNode`,  these methods are never used and can be cleaned up.

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/MdcAwareNodesComparator.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+package org.apache.ignite.spi.discovery.tcp.internal;
+
+import java.util.Comparator;
+
+/** Compares nodes using the Data Center id as a primary factor. */
+public class MdcAwareNodesComparator implements Comparator<TcpDiscoveryNode> {
+    /** */
+    @Override public int compare(TcpDiscoveryNode n1, TcpDiscoveryNode n2) {
+        String n1DcId = n1.dataCenterId() == null ? "" : n1.dataCenterId();
+        String n2DcId = n2.dataCenterId() == null ? "" : n2.dataCenterId();
```
Comment (alex-plekhanov) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/MdcAwareNodesComparator.java]: For multi-DC environment we have a check, that all `dataCenterId` is provided. We can ommit nullability check if comparator will be used only for multi-DC.

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/TcpDiscoveryNodesRing.java (hunk 2 above)
Comment (anton-vinogradov) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/TcpDiscoveryNodesRing.java]: MdcAwareNodesComparator guarantee elements to be non-equal since it compares dcId and if they are equal it compares nodes as usual.

```
if (res == 0)
   res = n1.compareTo(n2);
```

Comment (sergey-chugunov-1985) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/TcpDiscoveryNodesRing.java]: I agree that having a switch to turn this feature off and fall back to a manual management of ring structure is necessary.

There is always a chance that we miss an edge case during testing and a critical bug in discovery makes it to production. Having a flag as a way to get back to a suboptimal but working implementation is critical in such situation.

File: modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/MultiDataCenterRignTest.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.lang.IgniteInClosure;
+import org.apache.ignite.spi.discovery.DiscoverySpi;
+import org.apache.ignite.spi.discovery.tcp.messages.TcpDiscoveryAbstractMessage;
+import org.apache.ignite.spi.discovery.tcp.messages.TcpDiscoveryNodeAddedMessage;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.junit.Test;
+
+/** */
+public class MultiDataCenterRignTest extends GridCommonAbstractTest {
```
Comment (sergey-chugunov-1985) [inline modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/MultiDataCenterRignTest.java]:
```suggestion
public class MultiDataCenterRingTest extends GridCommonAbstractTest {
```

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java [hunk 6]
```diff
@@ -2733,42 +2699,15 @@ private static class PendingMessages implements Iterable<TcpDiscoveryAbstractMes
          */
         void add(TcpDiscoveryAbstractMessage msg) {
             msgs.add(new PendingMessage(msg));
-
-            while (msgs.size() > MAX) {
```
Comment (sergey-chugunov-1985) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java:2737]: As far as I can see, the old implementation have this logic to maintain internal queue size and not let it get close to MAX * 2 size right on add operation.

Now there is no such logic, we clean up internal collection on discard call itself.

Do we understand the logic behing this old behavior? What are situations when this logic starts working? Does new implementation perform better or worse in these situations?

This whole discard logic looks very important to me, and I want to make sure that this change doesn't break some corner case with lagging network which is not covered with our test base.

Comment (anton-vinogradov) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java:2737]: Old sollution (queue limiting) seems to be a buggy. In case of a queue overfill we just lose the consistency.
New solution guarantee the consistency and cleans the elements only when it's safe to clean.
Not sure we have deployments for 1K+ nodes where this will NOT happen now, but code becomes more stable, I think.

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/TcpDiscoveryNodesRing.java (hunk 2 above)
Comment (anton-vinogradov) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/TcpDiscoveryNodesRing.java]: Fixed with

```
private static final boolean mdcAwareRing = IgniteSystemProperties.getBoolean("MDC_AWARE_RING", true);
```

```
Collection<TcpDiscoveryNode> sorted;

if (mdcAwareRing) {
    sorted = new TreeSet<>(new MdcAwareNodesComparator());
    sorted.addAll(nodes);
}
else
    sorted = nodes;
```

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/MdcAwareNodesComparator.java (hunk 4 above)
Comment (anton-vinogradov) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/MdcAwareNodesComparator.java]: Removed the "" defaults.
